File: r-mos-frontend/src/types/maintenance.ts
export interface MaintenanceDraftStep {
  step_id: string
  title: string
  description: string
  model_targets?: string[]
  required_tools?: string[]
  preconditions?: string[]
}

export interface MaintenanceCitation {
  chunk_id?: string
  title: string
  score?: number
  source?: string
}

export interface MaintenanceDraftPayload {
  title: string
  maintenance_goal: string
  steps: MaintenanceDraftStep[]
  tools?: string[]
  citations?: MaintenanceCitation[]
  model_targets?: string[]
  review_notes?: string[]
  manifest_tree?: {
    robot_key?: string
    root_nodes?: string[]
    nodes?: Array<{
      id: string
      display_name?: string
      parent_id?: string | null
      children?: string[]
      source_paths?: string[]
      runtime_asset_paths?: string[]
      file_kinds?: string[]
    }>
  }
  manifest_mapping?: Record<
    string,
    {
      source_paths?: string[]
      runtime_asset_paths?: string[]
      file_kinds?: string[]
    }
  >
}

export interface MaintenanceDraftResponse {
  draft_id: string
  project_id: string
  request_id: string
  review_status: string
  draft: MaintenanceDraftPayload
  verdict_steps: Array<Record<string, unknown>>
  viewer_manifest: {
    robotId: string
    label?: string
    parts: string[]
    assets?: Array<{
      asset_id: string
      asset_type: string
      display_name?: string
      node_id: string
      path: string
      source_paths?: string[]
    }>
    structures?: Array<{
      path: string
      structure_type: string
      root_nodes?: string[]
    }>
    needs_review_nodes?: string[]
  }
  manifest_tree: NonNullable<MaintenanceDraftPayload['manifest_tree']>
  manifest_mapping: NonNullable<MaintenanceDraftPayload['manifest_mapping']>
  citations: MaintenanceCitation[]
}

export interface MaintenanceDraftCreateRequest {
  project_id?: string
  robot_key?: string
  maintenance_goal: string
  focus_area?: string
  request_id?: string
}

export interface MaintenanceDraftUpdateRequest {
  title?: string
  maintenance_goal?: string
  steps?: MaintenanceDraftStep[]
  tools?: string[]
  review_notes?: string[]
}

File: r-mos-backend/app/schemas/maintenance.py
from __future__ import annotations

from typing import Any, Optional

from pydantic import BaseModel, Field


class MaintenanceDraftCreateRequest(BaseModel):
    project_id: Optional[str] = None
    robot_key: Optional[str] = None
    maintenance_goal: str
    focus_area: Optional[str] = None
    request_id: Optional[str] = None


class MaintenanceDraftUpdateRequest(BaseModel):
    title: Optional[str] = None
    maintenance_goal: Optional[str] = None
    steps: Optional[list[dict[str, Any]]] = None
    tools: Optional[list[str]] = None
    review_notes: Optional[list[str]] = None


class DraftRejectRequest(BaseModel):
    reason: str = Field(..., min_length=1)


class MaintenanceDraftResponse(BaseModel):
    draft_id: str
    project_id: str
    request_id: str
    review_status: str
    draft: dict[str, Any]
    verdict_steps: list[dict[str, Any]]
    viewer_manifest: dict[str, Any]
    manifest_tree: dict[str, Any]
    manifest_mapping: dict[str, Any]
    citations: list[dict[str, Any]]
